### ISSUE TITLE
Contribution Proposal: ScatterND Implementation for PyTorch

### ISSUE THREAD

**Issue by boubezariali** [pytorch/pytorch#138502]
### 🚀 The feature, motivation and pitch

Hey everyone, I’m Ali, a software engineer from Nuro, and we’ve been working on migrating a lot of vision models from tensorflow to pytorch. One particularly critical operation for these vision models is tf.tensor_scatter_nd_update , and its reduction counterparts (e.g. tf.tensor_scatter_nd_min, tf.tensor_scatter_nd_max , etc).

We’ve implemented this operation in-house and have been using it successfully for a while now. There have been a few requests for an official pytorch implementation for this operation for a while now (e.g. [here](https://discuss.pytorch.org/t/how-to-implement-tf-scatter-nd-function-of-tensorflow-in-pytorch/18358), [here](https://discuss.pytorch.org/t/how-to-achieve-tf-scatter-nd-in-pytorch-project-for-copynet/4688), etc), so we’d like to offer to set this up.

It's a super robust implementation:
* Uses only native torch layers, and is fully statically shaped
* Supports batching with torch.vmap if desired
* Really performant when used with torch.compile & torch-xla

I've attached sample implementations in the 'Additional context section'.

### Alternatives

There are a few scatter_nd implementations scattered (pardon the pun) around the web but none seem to be perfect.
* https://github.com/rusty1s/pytorch_scatter, this one uses custom CUDA kernels which isn't easy to use or maintain
* https://gist.github.com/airalcorn2/c7846d6fcb58a30b25ea6d97e16fe025, this implementation has a bug
* https://gist.github.com/Ending2015a/b034ebbedc55fec1d8ec3b7230a95f1e, doesn't support batching, or have static shapes which is important for torch.compile/torch-xla

### Additional context

(sample implementation) https://gist.github.com/boubezariali/2e3d9650461f302a541235e33d7cded2
(simple test) https://gist.github.com/boubezariali/a4a22736a14414404b45b17f62da9c2d

We also have extensive testing implemented that we're willing to contribute.

**Comment by ezyang** [comment]
This seems like a useful wrapper! We're not sure how much general interest for it there is, and since it's pure Python it's easy for people to use if they need it. We'd probably have to figure out how much interest there is in the community for this to want to put it in.  Do you have benchmark results by any chance?

**Comment by boubezariali** [comment]
> This seems like a useful wrapper! We're not sure how much general interest for it there is, and since it's pure Python it's easy for people to use if they need it. We'd probably have to figure out how much interest there is in the community for this to want to put it in. Do you have benchmark results by any chance?

Given Google's pretty much announced the EOL of tensorflow, more and more projects will be looking to migrate tensorflow models to either pytorch or JAX. For Nuro, ScatterND ops were one of the most intimidating reasons not to switch to pytorch, so having this op as part of the core set could be helpful/encouraging to more groups to go with pytorch. Let me know if you need help gathering more information! 

Here are some benchmark results (in ms, done with `triton.testing.do_bench)`):
```
Torch CPU: 5.262639045715332
Torch CPU Compiled: 3.9849486351013184
Torch GPU: 1.3286479711532593
Torch GPU Compiled: 0.99783191108703613
TF CPU: 15.080506324768066
```
 I used very large inputs (300k indices/updates, with 100x100x100 output tensor). I couldn't figure out how to make Tensorflow use a GPU, but let me know if we really need that value. These speedups are consistent with what we noticed when moving from Tensorflow to torch!

**Comment by ezyang** [comment]
I think one problem will be to figure out where to put this, since in terms of "PyTorch" API-space, we would probably rather recommend people use index_put/scatter_reduce directly.

**Comment by boubezariali** [comment]
> I think one problem will be to figure out where to put this, since in terms of "PyTorch" API-space, we would probably rather recommend people use index_put/scatter_reduce directly.

This issue was a request for ScatterND to be added to the pytorch main API :) If you guys think that there is a better place to add this operation where it can still help lots of people, let me know! Happy to re-direct the contribution request.

**Comment by ezyang** [comment]
Yes. Can you concretely propose torch.XXXX that would implement this?

**Comment by boubezariali** [comment]
I've opened an issue in torchvision as well: https://github.com/pytorch/vision/issues/8703. Perhaps that will be a better fit.
